proptest/src/num/float_samplers.rs

```diff
@@ -0,0 +1,338 @@
+//-
+// Copyright 2022 Michael Wright
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Alternative uniform float samplers because the ones provided by the rand crate are prone
+//! to overflow. The samplers work by uniformly selecting from a set of equally spaced values in
+//! the interval and the included bounds. Selection is slighly biased towards the bounds.
+
+pub(crate) use self::f32::F32U;
+pub(crate) use self::f64::F64U;
+
+macro_rules! float_sampler {
+    ($typ: ident, $int_typ: ident, $wrapper: ident) => {
+        pub mod $typ {
+            use rand::prelude::*;
+            use rand::distributions::uniform::{
+                SampleBorrow, SampleUniform, Uniform, UniformSampler,
+            };
+
+            #[must_use]
+            // Returns the previous float value. In other words the greatest value representable
+            // as a float such that `last_float(a) < a`. `-0.` is treated as `0.`.
+            fn last_float(a: $typ) -> $typ {
```

Comment:
not sure about naming here, sounds more like `max_float_less_than_or_equal_to(a)` or we could copy swift's convention of `clamping` and call this `float_clamping(a)` or `clamping_float(a)` -- i think either of those work.

sauce https://developer.apple.com/documentation/swift/binaryinteger/init(clamping:)

Reply:
Ok, after reading through all the code more thoroughly i believe i understand the intent now. This is getting a float value that is the smallest bit increment less in magnitude from the input.

maybe `magnitude_decreased_by_bit(float)` or `minimally_smaller_float(float)`. i can't think of conventional language for this, but we should at least strive for higher clarity. given this is unconventional, i think i'm more in favor of the more explicit `magnitude_decreased_by_bit`.

Reply:
i think i'd actually prefer `minimally_smaller_float` since it conveys the intent/purpose that we are getting the immediate float that is smaller than this one, rather than the implementation -- that we're doing this by decrementing the bit value by 1. 

that said i dont have overly strong opinions.

Reply:
that's fair, i'll support that 👍

Reply:
appears to be common lingo https://www.google.com/search?client=firefox-b-1-d&q=next_down

there is a significant difference here though in that our impl does `next_up` for negatives, but the name at least is much closer to convention.